src/common/com/intellij/plugins/haxe/model/HaxeClassModel.java

```diff
@@ -39,6 +42,37 @@ public HaxeClassModel(HaxeClass haxeClass) {
     this.haxeClass = haxeClass;
   }
 
+  public HaxeProjectModel getProject() {
+    return HaxeProjectModel.fromElement(haxeClass);
+  }
+
+  public HaxeModuleModel getModule() {
+    return HaxeModuleModel.fromElement(haxeClass);
+  }
+
+  private HaxeFileModel file = null;
+
+  public HaxeFileModel getFile() {
+    if (file == null) file = new HaxeFileModel((HaxeFile)haxeClass.getContainingFile());
+    return file;
+  }
+
+  @Nullable
+  public FqInfo getFqInfo() {
+    PsiFile file = haxeClass.getContainingFile();
+    String directory = getProject().getPathToDirectory(file.getParent());
+    if (directory == null) return null;
+    String className = haxeClass.getName();
+    String fileName = HaxeFileNameUtils.removeExtension(file.getName());
+
+    return new FqInfo(directory.replace('/', '.'), fileName, className);
+  }
+
+  public HaxePackageModel getPackage() {
```

Comment:
Seems like the package should come from the package statement in the top of the file, not from reading the directory structure.  Also, if the package isn't specified, it belongs to the default namespace, in which case the directory structure is irrelevant.  The (old) haxe manual uses the word "should" when describing the relationship between packages and the directory structure, but it does not make the relationship mandatory: http://old.haxe.org/ref/packages.